### PR TITLE
[11.0][add][account_cryptocurrency]

### PR DESCRIPTION
**Pull request by JordiBForgeFlow**
Account Cryptocurrency
===================

This module provides basic support for the management of cryptocurrencies.

* Cryptocurrencies sare commonly considered property from a
  taxation standpoint.

* This means that every time that you receive or send cryptocurrencies, you
  are entering into a taxable transaction, and will need to report on the
  gain/loss resulting from the receipt/issue of the cryptocurrency.

* When you receive cryptocurrencies (in exchange for goods delivered,
  service rendered or because you mined it), that translates into taxable
  income to the recipient. The valuation should be based on the market value of
  the cryptocurrency at the time of receipt, and it can be provided by a
  third party (e.g. the exchange platform) or needs to be identified by
  you somehow.

* You have to retain the cost of the crytptocurrency for as long as you keep
  it.

* When you pay out cryptocurrencies (in exchange for goods or services
  purchased, or if you exchanged fpr another crypto or normal currency) this
  constitutes again a taxable transaction. In that case the gain/loss is
  computed as the difference between the cost of the cryptocurrency that
  you pay with, and the value of what you receive back, valued in your
  operating normal currency.

Installation
============

* Install the module 'account_invoicing' to be able to display the menus.

Configuration
=============

* For the user that will do the configuration, make sure to activate the
  option 'Show Full Accounting Features' in the user profile.

* Activate the Multi-Currency in 'Invoicing / Configuration /
  Settings / Multi-Currencies.

* Create the crypto currency in 'Invoicing / Configuration / Accounting /
  Currencies'. Complete the following fields:  'Inventoried', 'Valuation
  Method'.

* Add to the crypto currency an 'Inventory Account'. This Account must
  indicate in field 'Account Currency' the same currency.

* Create an Account Journal associated to the payments of the crypto
  currency, in 'Invoicing / Configuration / Accounting / Journals'. Indicate
  in the 'Currency' field the crypto currency. In the 'Default Credit Account'
  and 'Default Debit Account' you should indicate a new Account of type
  'Expenses', with a name similar to 'Crypto currency X Gain/Loss'. The new
  Account must also refer to the crypto currency account.

Additional remarks:

 * In case that your company currency is not EUR, make sure that the main
   company currency has no exchange rates. Go to 'Invoicing / Configuration /
   Accounting / Currencies' and for the company currency press the
   button 'Rates' and delete any rate listed.

Usage
=====

* If you have negotiated your invoices with customers / suppliers using
  crypto currencies, you can indicate them in the corresponding invoice.

* When you receive/send payments you can indicate the crypto currency that
  you want to use for the payment. Once the payment has been posted you will
  be able to navigate to the associated currency moves.

* You cannot make internal transfers using cryptocurrencies.

### CODE REVIEW

**Comment by JordiBForgeFlow** [comment]
@mart-e  this might me interesting to you.

File: account_cryptocurrency/README.rst
Comment (elicoidal) [inline account_cryptocurrency/README.rst]: target is missing

File: account_cryptocurrency/README.rst
Comment (elicoidal) [inline account_cryptocurrency/README.rst]: sare->are

File: account_cryptocurrency/README.rst
Comment (elicoidal) [inline account_cryptocurrency/README.rst]: you might want to use enumerated list

File: account_cryptocurrency/README.rst
Comment (elicoidal) [inline account_cryptocurrency/README.rst]: Can you add runbot

File: account_cryptocurrency/README.rst
Comment (elicoidal) [inline account_cryptocurrency/README.rst]: fix the URL

File: account_cryptocurrency/__manifest__.py
Comment (elicoidal) [inline account_cryptocurrency/__manifest__.py]: no date?

File: account_cryptocurrency/__manifest__.py
Comment (elicoidal) [inline account_cryptocurrency/__manifest__.py]: fix the URL

File: account_cryptocurrency/data/res_currency_move_sequence.xml
Comment (elicoidal) [inline account_cryptocurrency/data/res_currency_move_sequence.xml]: merge with odoo

File: account_cryptocurrency/data/res_currency_move_sequence.xml
Comment (elicoidal) [inline account_cryptocurrency/data/res_currency_move_sequence.xml:1]: no need encoding

File: account_cryptocurrency/data/res_currency_move_sequence.xml
Comment (elicoidal) [inline account_cryptocurrency/data/res_currency_move_sequence.xml]: very long sequence number. CM is good enough no?

File: account_cryptocurrency/models/res_currency.py
Comment (elicoidal) [inline account_cryptocurrency/models/res_currency.py]: First in First Out or First-in-First-Out

File: account_cryptocurrency/security/account_cryptocurrency_security.xml
Comment (elicoidal) [inline account_cryptocurrency/security/account_cryptocurrency_security.xml]: merge with Odoo

File: account_cryptocurrency/security/account_cryptocurrency_security.xml
Comment (elicoidal) [inline account_cryptocurrency/security/account_cryptocurrency_security.xml]: Can you improve the name, too "cryptic"

File: account_cryptocurrency/security/account_cryptocurrency_security.xml
Comment (elicoidal) [inline account_cryptocurrency/security/account_cryptocurrency_security.xml]: Sometimes you use Crypto currency, sometimes cryptocurrency

File: setup/account_cryptocurrency/odoo/addons/account_cryptocurrency
Comment (elicoidal) [inline setup/account_cryptocurrency/odoo/addons/account_cryptocurrency:1]: remove these files

**Review by elicoidal — CHANGES_REQUESTED** [review]
Good job!
Do you think the module needs some demo data?

File: account_cryptocurrency/README.rst
Comment (JordiBForgeFlow) [reply account_cryptocurrency/README.rst]: Done

File: account_cryptocurrency/__manifest__.py
Comment (JordiBForgeFlow) [reply account_cryptocurrency/__manifest__.py]: Done.

File: account_cryptocurrency/data/res_currency_move_sequence.xml
Comment (JordiBForgeFlow) [reply account_cryptocurrency/data/res_currency_move_sequence.xml:1]: Not sure about this.

File: setup/account_cryptocurrency/odoo/addons/account_cryptocurrency
Comment (JordiBForgeFlow) [reply setup/account_cryptocurrency/odoo/addons/account_cryptocurrency:1]: Why? This is a symlink and it is required to be in the setup folder for the pip install method.

**Comment by JordiBForgeFlow** [comment]
@elicoidal many thanks for your review! I have attended your comments.

File: setup/account_cryptocurrency/odoo/addons/account_cryptocurrency
Comment (elicoidal) [reply setup/account_cryptocurrency/odoo/addons/account_cryptocurrency:1]: I thought the set up folder was created automatically

File: account_cryptocurrency/data/res_currency_move_sequence.xml
Comment (elicoidal) [reply account_cryptocurrency/data/res_currency_move_sequence.xml:1]: @pedrobaeza do we have a guideline for that?

File: account_cryptocurrency/data/res_currency_move_sequence.xml
Comment (elicoidal) [inline account_cryptocurrency/data/res_currency_move_sequence.xml]: I personally think that sequence should be short and without "/" like CM180001 but non blocking

File: account_cryptocurrency/models/account_account.py
Comment (elicoidal) [inline account_cryptocurrency/models/account_account.py]: remove comma

**Review by elicoidal — APPROVED** [review]
non blocking details

File: setup/account_cryptocurrency/odoo/addons/account_cryptocurrency
Comment (JordiBForgeFlow) [reply setup/account_cryptocurrency/odoo/addons/account_cryptocurrency:1]: It is, but if you need to pip install the module before it is merged, you need to add it manually.

**Comment by JordiBForgeFlow** [comment]
@lreficent @mreficent reviews welcome!

File: account_cryptocurrency/README.rst
Comment (moylop260) [reply account_cryptocurrency/README.rst]: The new [README template](https://raw.githubusercontent.com/OCA/maintainer-tools/master/template/module/README.rst) uses:
```rst
.. image:: https://img.shields.io/badge/license-AGPL--3-blue.png
   :target: https://www.gnu.org/licenses/agpl
   :alt: License: AGPL-3
```

File: account_cryptocurrency/__manifest__.py
Comment (moylop260) [reply account_cryptocurrency/__manifest__.py]: Use `https://github.com/OCA/currency` instead

File: account_cryptocurrency/models/account_account.py
Comment (moylop260) [inline account_cryptocurrency/models/account_account.py]: If the filtered is truly then you can use a string with the name of the field `self.filtered('currency_id')`

File: account_cryptocurrency/models/account_bank_statement.py
Comment (moylop260) [inline account_cryptocurrency/models/account_bank_statement.py]: What about?
```python
if self.filtered('journal_id.currency_id.inventoried'):
   raise...
```

File: account_cryptocurrency/models/account_payment.py
Comment (moylop260) [inline account_cryptocurrency/models/account_payment.py]: What about `if self.currency_id and self.currency_id.inventoried:`?

File: account_cryptocurrency/models/account_payment.py
Comment (moylop260) [inline account_cryptocurrency/models/account_payment.py:69]: What about 
```python
if self.filtered(lambda rec: rec.currency_id.inventoried and rec.payment_type == 'transfer'):
    raise ...
```

File: account_cryptocurrency/models/res_currency.py
Comment (moylop260) [inline account_cryptocurrency/models/res_currency.py:20]: Remove string because is the same that technical name.
https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#fields

```md
If the technical name of the field (the variable name) is the same to the string of the label, don't put string parameter for new API fields, because it's automatically taken. If your variable name contains "_" in the name, they are converted to spaces when creating the automatic string and each word is capitalized. 
```

File: account_cryptocurrency/models/res_currency_move.py
Comment (moylop260) [inline account_cryptocurrency/models/res_currency_move.py:54]: A `compute` field is `readonly` by default maybe you could delete this parameter

File: account_cryptocurrency/models/res_currency_move.py
Comment (moylop260) [inline account_cryptocurrency/models/res_currency_move.py:86]: Shouldn't be used `rec` instead of `self` here and lines below?

File: account_cryptocurrency/models/res_currency_move.py
Comment (moylop260) [inline account_cryptocurrency/models/res_currency_move.py]: FYI for python3.5 you can use `return super().create(vals)`

File: account_cryptocurrency/models/res_currency_move.py
Comment (moylop260) [inline account_cryptocurrency/models/res_currency_move.py]: What about `for candidate in candidates.filtered('remaining_qty'):`?
Using this way you reduce the cyclomatic complexity and then reduces the qty of tests to cover this line

File: account_cryptocurrency/models/res_currency_move.py
Comment (moylop260) [inline account_cryptocurrency/models/res_currency_move.py]: If you are using floating amounts. Shouldn't use `currency_id.is_zero(amount)` method?

File: account_cryptocurrency/models/res_currency_move.py
Comment (moylop260) [inline account_cryptocurrency/models/res_currency_move.py]: What about `if self.mapped('move_line_ids'):`?

File: account_cryptocurrency/models/res_currency_move_line.py
Comment (moylop260) [inline account_cryptocurrency/models/res_currency_move_line.py]: same case self.mapped('account_move_ids') here

**Review by moylop260 — APPROVED** [review]
Congrats and thanks!
This module is very very fine!

Just a few small comments (non blocking)

**Comment by JordiBForgeFlow** [comment]
Thanks @moylop260 I have attended your comments. See also a real case of purchase of crypto and management using this and other OCA modules explained here: https://docs.google.com/document/d/1IKKenj2d7RICrx2DcMIOCvar16d7pG8bCmhUcYU95yE/edit

My plan is to update the document with more complex scenarios.

Still I have pending another module account_cryptocurrency_landed_cost that should allow to allocate the various fees associated to crypto currency purchases to the coin base cost, very similar to what stock_landed_cost does with stock valuation in real time.

**Review by LoisRForgeFlow — APPROVED** [review]
LGTM :+1:

File: account_cryptocurrency/data/res_currency_move_sequence.xml
Comment (MiquelRForgeFlow) [reply account_cryptocurrency/data/res_currency_move_sequence.xml:1]: https://github.com/OCA/maintainer-tools/blob/master/template/module/data/model_name_data.xml#L1

**Pull request merged**
